### PR TITLE
Lra 321 avm monitoring dashboard

### PR DESCRIPTION
**Pull request by britaumich**
The branch has a dashboard with filters and a one-room show page.

To test please run 'bin/rails db:seed' to populate the database

To test the attention-needed filter run the "DeviceState.create(key: "Error State", value: "true", device_id: 6)" command in the rails console to change a Projector's state, then reload the dashboard.

### CODE REVIEW

File: app/javascript/controllers/autosubmit_controller.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { Controller } from "@hotwired/stimulus"
+export default class extends Controller {
+  static targets = ['form', 'status', 'sidebar', 'min_capacity', 'max_capacity', 'capacity_error']
+
+  connect() {
+    console.log("Hello World!")
```
Comment (anantaUM) [inline app/javascript/controllers/autosubmit_controller.js:6]: Please update this message such that it is more relevant to this application.

File: app/javascript/controllers/autosubmit_controller.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  }
+
+  search() {
+    clearTimeout(this.timeout)
+
```
Comment (anantaUM) [inline app/javascript/controllers/autosubmit_controller.js:11]: Remove extra space on line 11

File: app/views/rooms/_room.html.erb [hunk 3]
```diff
@@ -1,32 +1,11 @@
...
-    <strong class="block font-medium mb-1">Room type:</strong>
-    <%= room.room_type %>
-  </p>
-
-  <% if action_name != "show" %>
-    <%= link_to "Show this room", room, class: "rounded-lg py-3 px-5 bg-gray-100 inline-block font-medium" %>
-    <%= link_to 'Edit this room', edit_room_path(room), class: "rounded-lg py-3 ml-2 px-5 bg-gray-100 inline-block font-medium" %>
-    <hr class="mt-6">
-  <% end %>
+
```
Comment (anantaUM) [inline app/views/rooms/_room.html.erb:2]: Please remove extra space on line 2.

File: app/views/rooms/_room_assets.html.erb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+<div class="text-xs px-2">
+
```
Comment (anantaUM) [inline app/views/rooms/_room_assets.html.erb:2]: Please remove extra space on line 2.

File: app/views/rooms/_room_assets.html.erb [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<div class="text-xs px-2">
+  
+  <% devices.all.each do |device| %>
+    <span class="font-bold"><%= device.name %></span>
+    <%# device.device_states.each do |state| %>
```
Comment (anantaUM) [inline app/views/rooms/_room_assets.html.erb:5]: Please remove this comment code if not needed.

File: app/views/rooms/_room_card.html.erb [hunk 6]
```diff
@@ -0,0 +1,29 @@
+<%= turbo_frame_tag dom_id(room) do  %>
+  <% @attention = false %>
+  <h2 class="sr-only">room</h2>
+  <div id="room" class="mt-2 mb-3 px-2 sm:px-0">
+    <div>
+      <div class="text-xs px-2 font-bold hover:underline">
+        <%= link_to room.facility_id, room, data: { "turbo-frame": "_top" } %>
+
```
Comment (anantaUM) [inline app/views/rooms/_room_card.html.erb:8]: Please remove extra space on line 8, 26 and 28.

File: app/views/rooms/_room_state.html.erb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+<div class="text-xs px-2">
+  <%# device.device_states.all.each do |state| %>
```
Comment (anantaUM) [inline app/views/rooms/_room_state.html.erb:2]: Please remove this commented code if not needed.

File: app/views/rooms/index.html.erb [hunk 8]
```diff
@@ -1,14 +1,71 @@
...
-
-  <div class="flex justify-between items-center">
-    <h1 class="font-bold text-4xl">Rooms</h1>
-    <%= link_to 'New room', new_room_path, class: "rounded-lg py-3 px-5 bg-blue-600 text-white block font-medium" %>
+<div class="container px-1 mb-20 pl-4" data-controller='autosubmit'>
+  <h1 class="text-xl font-semibold text-gray-900 p-2">Rooms with Crestron devices</h1>
+  <div class="grid grid-cols-2 pb-2">
+    <div>
+      
+    </div>
```
Comment (anantaUM) [inline app/views/rooms/index.html.erb:6]: Looks like nothing is happening in this <div> element. Please remove it.

File: app/views/rooms/index.html.erb [hunk 9]
```diff
@@ -1,14 +1,71 @@
...
+            </div>
+          </div>
+        </div>
+        <div class="">
+          <%= render 'listing' %>
+        </div>
+      </div>
+    <% end %>
   </div>
+
```
Comment (anantaUM) [inline app/views/rooms/index.html.erb:69]: Please remove extra space on line 51 and 69.

File: app/controllers/rooms_controller.rb [hunk 10]
```diff
@@ -1,9 +1,39 @@
...
+                throw :attention 
+              end
+            end
+          end
+        end
+        
+      end
+      @rooms = attention_rooms
+    end
+
```
Comment (anantaUM) [inline app/controllers/rooms_controller.rb:30]: Please remove extra space on line 9 and 26.

File: app/views/layouts/_header.html.erb [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+        </div>
+        <div class="hidden sm:ml-6 sm:flex sm:space-x-4 pl-2">
+          <div class="border-transparent text-blue-50 hover:border-blue-300 hover:text-white inline-flex items-center pl-1 pt-1 border-b-2 text-base font-normal">
+            <%= link_to "Find a Room", rooms_path %>
+          </div>
+        </div>
+      </div>
+    </div>
+  </div>
+
```
Comment (anantaUM) [inline app/views/layouts/_header.html.erb:25]: Please remove this extra space.

File: app/helpers/application_helper.rb [hunk 12]
```diff
@@ -1,2 +1,119 @@
...
+        DeviceState.where(device_id: device.id).select(:key, :value, 'MAX(created_at)').group(:key).each do |state|
+          if state_need_attention?(state)
+            att = true
+            throw :attention 
+          end
+        end
+      end
+    end
+    return att
+
```
Comment (anantaUM) [inline app/helpers/application_helper.rb:35]: Please remove extra space on line 35 and 118.

**Review by anantaUM — COMMENTED** [review]
Added comments.

File: app/helpers/application_helper.rb (hunk 12 above)
Comment (britaumich) [reply app/helpers/application_helper.rb:35]: I don't think space on line 118 is extra. I agree about space on line 38

File: app/javascript/controllers/autosubmit_controller.js (hunk 1 above)
Comment (britaumich) [reply app/javascript/controllers/autosubmit_controller.js:6]: I'll remove the message. It has no special meaning. Actually, I'll remove the whole method. I needed it only to check that the controller worked.

File: app/views/rooms/_audio.html.erb [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+      WL Mic - <%= wl_mic(room) %>
+    </div>
+    <div>
+     Mic Volume - <%= wl_mic_volume(room) %>
+    </div>
+    <div>
+     Source Volume - <%= source_volume(room) %>
+    </div>
+  </div>
+
```
[the remark this answers is not in the document]
Comment (britaumich) [reply app/views/rooms/_audio.html.erb:21]: I will

File: app/views/rooms/_listing.html.erb [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+    <span class="text-muted" data-autosubmit-target="status"></span>
+  </div>
+  <div class="pl-2 font-semibold">
+    All rooms
+  </div>
+  <div class="">
+    <ul role="list" class="grid grid-cols-2 gap-4 sm:grid-cols-2 lg:grid-cols-4 m-4">
+      <% @rooms.each do |room| %>
+        <li class="col-span-1 bg-white rounded-lg shadow-md divide-y divide-gray-400 px-2 grid place-items-stretch bg-white border border-gray-300 rounded-lg">
+          <%#= render 'room_card', { data: room_crestron['data'], room: Room.find_by(facility_id: room_crestron['facility_id']), attention: room_crestron['attention'] } %>
```
[the remark this answers is not in the document]
Comment (britaumich) [reply app/views/rooms/_listing.html.erb:12]: I will

File: app/views/rooms/_room_assets.html.erb (hunk 4 above)
Comment (britaumich) [reply app/views/rooms/_room_assets.html.erb:2]: thank you for being so patient - I will do that

File: app/views/rooms/_room_assets.html.erb (hunk 5 above)
Comment (britaumich) [reply app/views/rooms/_room_assets.html.erb:5]: of course

File: app/views/rooms/_room_card.html.erb (hunk 6 above)
Comment (britaumich) [reply app/views/rooms/_room_card.html.erb:8]: I will

File: app/views/rooms/_room_state.html.erb (hunk 7 above)
Comment (britaumich) [reply app/views/rooms/_room_state.html.erb:2]: yes

File: app/views/rooms/index.html.erb (hunk 8 above)
Comment (britaumich) [reply app/views/rooms/index.html.erb:6]: I guess something was there ... - will remove

File: app/views/rooms/index.html.erb (hunk 9 above)
Comment (britaumich) [reply app/views/rooms/index.html.erb:69]: yes

File: app/javascript/controllers/autosubmit_controller.js (hunk 2 above)
Comment (britaumich) [reply app/javascript/controllers/autosubmit_controller.js:11]: I like this space, but I'll remove it for you

File: app/controllers/rooms_controller.rb (hunk 10 above)
Comment (britaumich) [reply app/controllers/rooms_controller.rb:30]: will do

**Comment by britaumich** [comment]
I removed extra spaces and comments. Fixed a file path. But I don't know what do you want me to fix about application_helper.

File: app/helpers/application_helper.rb [hunk 15]
```diff
@@ -1,2 +1,119 @@
 module ApplicationHelper
+
```
[the remark this answers is not in the document]
Comment (britaumich) [reply app/helpers/application_helper.rb:2]: done last week

**Review by anantaUM — APPROVED** [review]

**Pull request merged**